libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,221 @@
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, !adc_mode);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+  ADC_ITConfig(ADC1, ADC_IER_EOSEQIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.sequence = 0;
+
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+
+  // Configure internal reference channel to run by default for voltage conversions
+  adc_set_channel(ADC_CHANNEL_REF, ENABLE);
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= ~(1 << adc_channel);
+  }
+
+  // Keep internal channels enabled only when set
+  switch (adc_channel) {
+    case ADC_CHANNEL_BAT:
+      ADC_VbatCmd(new_state);
+      break;
+
+    case ADC_CHANNEL_REF:
+      ADC_VrefintCmd(new_state);
+      break;
+
+    case ADC_CHANNEL_TEMP:
+      ADC_TempSensorCmd(new_state);
+      break;
+  }
+
+  s_adc_status.sequence = ADC1->CHSELR;
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(ADCChannel adc_channel, ADCCallback callback, void *context) {
+  // Returns invalid if the given address is not connected to an ADC channel
+  StatusCode invalid = prv_channel_valid(adc_channel);
+
+  if (invalid) {
+    return invalid;
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
+  s_adc_interrupts[adc_channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_read_raw(ADCChannel adc_channel, uint16_t *reading) {
+  StatusCode invalid = prv_channel_valid(adc_channel);
+
+  if (invalid) {
+    return invalid;
+  }
+
+  if (!s_adc_status.continuous) {
+    ADC_StartOfConversion(ADC1);
+    while (ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) { }
+  }
+
+  *reading = s_adc_interrupts[adc_channel].reading;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_read_converted(ADCChannel adc_channel, uint16_t *reading) {
+  StatusCode invalid = prv_channel_valid(adc_channel);
+
+  if (invalid) {
+    return invalid;
+  }
+
+  uint16_t adc_reading, internal_reading;
+
+  switch (adc_channel) {
+    case ADC_CHANNEL_TEMP:
+      internal_reading = s_adc_interrupts[ADC_CHANNEL_TEMP].reading;
+      adc_reading = prv_get_temp(internal_reading);
+      break;
+
+    case ADC_CHANNEL_REF:
+      internal_reading = s_adc_interrupts[ADC_CHANNEL_REF].reading;
+      adc_reading = prv_get_vdda(internal_reading);
+      break;
+
+    case ADC_CHANNEL_BAT:
+      adc_reading = s_adc_interrupts[ADC_CHANNEL_BAT].reading * 2;
+      break;
+
+    default:
+      adc_read_raw(adc_channel, reading);
```

Comment:
Would recommend using your own function (`adc_read_converted`) instead of rewriting it in a switch case